src/engraving/libmscore/segment.cpp

```diff
@@ -508,7 +508,9 @@ void Segment::insertStaff(staff_idx_t staff)
 
     for (EngravingItem* e : _annotations) {
         staff_idx_t staffIdx = e->staffIdx();
-        if (staffIdx >= staff && !e->isTopSystemObject()) {
+        bool keepSystemObjectOnTop = e->systemFlag() && staffIdx == 0
+                                     && !(staff + 1 < score()->nstaves() && score()->staff(staff + 1)->isSystemObjectStaff());
```

Comment:
I have already seen similar conditions in this PR several times. All of them should be simplified, comments should be added to them. It is very difficult to read/debug them in their current state. In addition, they can potentially be moved to a function if they are duplicates

Reply:
Yeah, all those conditions are totally horrible. Basically, they try to account for all the different combinations that you can delete/reinsert staves and how to deal with the systemObjects in each case (sometime the object should be deleted with the staff, sometimes not, sometimes it must be re-inserted with the staff, sometimes not, etc etc...). And each of those cases is slightly different, so it's hard to extract this logic into a re-usable piece. I'll do my best to tidy it up and make it as clear as possible